### PR TITLE
[WIP] Updating require_hardcoding_lib AST to report an offense when a relative path is used to lib/ (and, not specifically focus on require calls)

### PR DESCRIPTION
**Pull request by samyak-jn**
Hi,

This is with regards to issue #35 for the same I have tweaked the AST so that it checks the usage of File.expand_path and File.dirname used with lib/.
I am not so much proficient with ruby, so please correct me if I'm wrong, thanks! :)

Once, the AST is finalized we can discuss more renaming the cop and etcetera.

### CODE REVIEW

File: spec/rubocop/cop/packaging/require_hardcoding_lib_spec.rb [hunk 1]
```diff
@@ -5,45 +5,9 @@
...
-      RUBY
-
-      expect_correction(<<~RUBY)
-        $:.unshift('../../lib')
-        require "foo/bar"
-      RUBY
-    end
-  end
-
-  context "when `require` call uses File#expand_path method with __FILE__" do
```
Comment (utkarsh2102) [inline spec/rubocop/cop/packaging/require_hardcoding_lib_spec.rb:44]: I expect these tests to be still working, really. Might need some fixing but in the end, it should work, too (even as-is, from a quick POV). Because we're essentially just generalizing the entire set of calls and the above tests are really just a subset of those. So I expect them to work + I expect new tests to be added (the generalized one, mentioned in the issue, for example).

File: spec/rubocop/cop/packaging/require_hardcoding_lib_spec.rb [hunk 2]
```diff
@@ -137,64 +101,9 @@
...
-
-    it "does not register an offense" do
-      expect_no_offenses(<<~RUBY, filename)
-        #{source}
-      RUBY
-    end
-  end
-
-  context "when `require` call uses File#expand_path method with __FILE__ but lies inside lib/" do
+  context "when use of File#expand_path method with __FILE__ but lies inside lib/" do
```
Comment (utkarsh2102) [inline spec/rubocop/cop/packaging/require_hardcoding_lib_spec.rb:104]: Same here.

File: lib/rubocop/cop/packaging/require_hardcoding_lib.rb [hunk 3]
```diff
@@ -65,7 +64,7 @@ def on_new_investigation
         # More about the `#on_send` method can be found here:
         # https://github.com/rubocop-hq/rubocop-ast/blob/08d0f49a47af1e9a30a6d8f67533ba793c843d67/lib/rubocop/ast/traversal.rb#L112
         def on_send(node)
-          return unless require?(node)
+          return unless File?(node)
```
Comment (utkarsh2102) [inline lib/rubocop/cop/packaging/require_hardcoding_lib.rb:67]: We need a different name here, `File` is not very clear as to what it implies or means. Also preferably something that's not capitalized. Hope that makes sense?

**Review by utkarsh2102 — CHANGES_REQUESTED** [review]
Hi @samyak-jn,

Thanks for working on this but I am not sure if this is what we need? We should essentially generalize the AST to catch all the "bad" calls and "require" ones were just a subset of those. I don't expect someone to go through the AST in detail unless we have the existing tests passing + we've added more generalized tests, which are passing as well. The AST should then "speak for itself" and maybe someone could then do the final edge-case testing and stuff and do a finish-up of the review/merge. Does that make sense? 😅 

Let me know if we're not on the same page or something or if I am simply missing anything?

**Comment by samyak-jn** [comment]
Hi @utkarsh2102,
 
> Thanks for working on this but I am not sure if this is what we need? We should essentially generalize the AST to catch all the "bad" calls and "require" ones were just a subset of those.
> Let me know if we're not on the same page or something or if I am simply missing anything?

Thanks for quite a verbose explanation, and yes I agree with your point on this. 
Before digging deeper I wanted to make sure if I'm proceeding in the right direction, currently, the cop checks just for the `require` calls and we want this cop to just not specifically check for former but also to check for the relative path being used to lib/ (which basically focus on the implementation of File.dirname and File.expand_path methods), right?

For the very same here's the AST below could you please verify them?

```

        def_node_matcher :require?, <<~PATTERN
          {(send nil? :require (str #falls_in_lib?))
           (send nil? :require (send (const nil? :File) :expand_path (str #falls_in_lib?) (send nil? :__dir__)))
           (send nil? :require (send (const nil? :File) :expand_path (str #falls_in_lib_using_file?) (str _)))
           (send nil? :require (send (send (const nil? :File) :dirname {(str _) (send nil? _)}) :+ (str #falls_in_lib_with_file_dirname_plus_str?)))
           (send nil? :require (dstr (begin (send (const nil? :File) :dirname {(str _) (send nil? _)})) (str #falls_in_lib_with_file_dirname_plus_str?)))
           (send (const nil? :File) :expand_path (str #falls_in_lib?) (send nil? :__dir__))
           (send (const nil? :File) :expand_path (str #falls_in_lib_using_file?) (str _))
           (send (send (const nil? :File) :dirname {(str _) (send nil? _)}) :+ (str #falls_in_lib_with_file_dirname_plus_str?))
           (dstr (begin (send (const nil? :File) :dirname {(str _) (send nil? _)})) (str #falls_in_lib_with_file_dirname_plus_str?))}
```

Thanks for taking your out time for this :)

File: lib/rubocop/cop/packaging/require_hardcoding_lib.rb (hunk 3 above)
Comment (samyak-jn) [reply lib/rubocop/cop/packaging/require_hardcoding_lib.rb:67]: Yes, of course, thanks for PTO. :)

**Comment by utkarsh2102** [comment]
Oh no, I don't think we want `:require` in the AST. The AST could be generic as what you have in the PR atm, but that AST should also be able to detect the require calls. So basically, from "require"-specific AST, we want to move towards something generic that'd catch all sorts of calls that use relative path, **including** the require calls themselves, but the "require" keyword shouldn't be an offense anymore.

**Comment by utkarsh2102** [comment]
+ I guess the CI should be green, that's the very first thing I'd like to see, heh.

**Comment by mvz** [comment]
I'm not sure this direction is correct. `require` has specific logic for finding the required file in the library path, but this doesn't apply for any other use of relative paths.
